File: Haskell/Implementations/Command/Operator.hs
module Operator
    ( human1
    , human2
    ) where

import Command (Command(..), dispatch)
import ConsoleMailServerT (ConsoleMailServerT(..))
import Email (Email(..), emptyEmail)
import EmailEditorT (EmailEditorT(..))
import MailServer (SendEmail, SignEmail)
import StateT (StateT(..))
import User (User, mkUser, newCommand, runCommandsAsUser)

sendEditedEmail :: String -> Command
sendEditedEmail = SendEmail

signEmailWithServer :: Command
signEmailWithServer = SignEmail

writeSubject :: String -> Command
writeSubject = UpdateText (\s email -> email { subject = s })

writeBody :: String -> Command
writeBody = UpdateText (\s email -> email { body = s })

runCommands :: ConsoleMailServerT (EmailEditorT IO) a -> IO ()
runCommands p = do
    _ <-
        (\p' -> runStateT p' emptyEmail)
        $ runEmailEditorT
        $ runConsoleMailServerT p
    pure ()

human1 :: IO ()
human1 = do
    _ <- runCommands (runCommandsAsUser user)
    pure ()
  where
    user =
        newCommand sendTheEmail
            . newCommand signTheEmail
            . newCommand writeTheSubject
            . newCommand changeBody
            . newCommand writeInitialBody
            $ mkUser "Human 1"
    writeInitialBody = writeBody "Foo bar baz quux quack"
    changeBody       = writeBody "Eleven point eight inches"
    writeTheSubject  = writeSubject "Foobar"
    signTheEmail     = signEmailWithServer
    sendTheEmail     = sendEditedEmail "foo.bar@quack.quux"

human2 :: IO ()
human2 = do
    _ <- runCommands (runCommandsAsUser firstEmailUser)
    putStrLn "-----"
    _ <- runCommands (runCommandsAsUser secondEmailUser)
    pure ()
  where
    user                     = mkUser "Human 2"
    recipient                = "grace.hopper@us.navy"
    writeInitialBody         = writeBody "Hello"
    writeMoreBody            = writeBody "Hello there"
    writeEvenMoreBody        = writeBody "Hello th"
    writeFinalBody           = writeBody "Hello, dear reader"
    writeTheSubject          = writeSubject "To hello"
    correctTheSubject        = Macro [writeSubject "", writeSubject "To reader"]
    sendTheEmail             = sendEditedEmail recipient
    sendEmailAgainByAccident = sendEditedEmail recipient
    firstEmailUser =
        newCommand sendTheEmail
            . newCommand sendTheEmail
            . newCommand correctTheSubject
            . newCommand writeTheSubject
            . newCommand writeFinalBody
            . newCommand writeEvenMoreBody
            . newCommand writeMoreBody
            . newCommand writeInitialBody
            $ user

    writeApologeticBody    = writeBody "Sorry for sending it twice"
    writeApologeticSubject = writeSubject "RE: To reader"
    sendApologeticEmail    = sendEditedEmail recipient
    secondEmailUser =
        newCommand sendApologeticEmail
            . newCommand writeApologeticSubject
            . newCommand writeApologeticBody
            $ user

File: Java/Implementations/Command/Operator.java
import java.util.function.Supplier;

/**
 * A human operating the system in question.
 */
public class Operator {
	private EmailEditor editor;
	private MailServer mailServer;

	public Operator(MailServer mailServer, EmailEditor editor) {
		this.mailServer = mailServer;
		this.editor = editor;
	}

	public Command sendEditedEmail(String recipient) {
		return new SendEmailCommand(this.mailServer, this.editor, recipient);
	}

	public Command signEmailWithServer() {
		return new SignEmailCommand(this.mailServer, this.editor);
	}

	public Command writeSubject(String subject) {
		return new TextCommand(t -> this.editor.subject(t), subject);
	}

	public Command writeBody(String body) {
		return new TextCommand(t -> this.editor.body(t), body);
	}

	public static void human1(MailServer mailServer, EmailEditor editor) {
		User user = new User("Human 1");
		Operator human = new Operator(mailServer, editor);

		Command[] commands = new Command[]{
			human.writeBody("Foo bar baz quux quack"),
			human.writeBody("Eleven point eight inches"),
			human.writeSubject("Foobar"),
			human.signEmailWithServer(),
			human.sendEditedEmail("foo.bar@quack.quux")
		};

		for (Command command : commands) {
			user.newCommand(command);
		}

		user.runCommandsAsUser();
	}

	public static void human2(MailServer mailServer, Supplier<EmailEditor> editorFactory) {
		User user = new User("Human 2");
		EmailEditor editor = editorFactory.get();
		Operator human = new Operator(mailServer, editor);
		String recipient = "grace.hopper@us.navy";

		Command[] commands = new Command[]{
			human.writeBody("Hello"),
			human.writeBody("Hello there"),
			human.writeBody("Hello th"),
			human.writeBody("Hello, dear reader"),
			human.writeSubject("To hello"),
			new MacroCommand()
				.add(human.writeSubject(""))
				.add(human.writeSubject("To reader")),
			human.sendEditedEmail(recipient),
			human.sendEditedEmail(recipient),
		};

		for (Command command : commands) {
			user.newCommand(command);
		}

		user.runCommandsAsUser();

		System.out.println("-----");

		editor = editorFactory.get();

		// 63

		Command[] commands = new Command[]{
			human.writeBody("Sorry for sending it twice"),
			human.writeSubject("RE: To reader"),
			human.sendEditedEmail(recipient)
		};

		for (Command command : commands) {
			user.newCommand(command);
		}

		user.runCommandsAsUser();
	}
}
